### PR TITLE
[SIW-984, SIW-989, SIW-982, SIW-987, SIW-985, SIW-993, SIW-990] Infrastructure setup

### PR DESCRIPTION
**Pull request by balanza**
Setup the initial architecture for the cloud resources of the Wallet Provider as per the following schema:

alt text

All the resources are related to the `io-p-itn-wallet-rg-01` resource group and defined into the Italy North Azure region.

CDN will respond at the custom domain [wallet.io.pagopa.it](https://wallet.io.pagopa.it) and will serve static assets such as [.well-known/openid-federation](https://wallet.io.pagopa.it/.well-known/openid-federation)

A key vault is defined to hold signing keys for signed assets such as Wallet Instance Attestation. The function app is granted read access to it.

> ⚠️ The terraform project is already applied and the resources are already available on the `PROD-IO` subscription

### CODE REVIEW

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  name                = "${var.project}-cosmos-wallet"
+  resource_group_name = var.resource_group_name
+  location            = var.location
+  domain              = "WALLET"
+
+  offer_type = "Standard"
+  kind       = "GlobalDocumentDB"
+
+  main_geo_location_zone_redundant = false
```
Comment (pasqualedevita) [inline infra/resources/_modules/cosmos/cosmos_account.tf:12]:
```suggestion
  main_geo_location_zone_redundant = true
```

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    max_staleness_prefix    = null
+  }
+
+  main_geo_location_location = var.location
+
+  additional_geo_locations = [
+    {
+      location          = var.secondary_location
+      failover_priority = 1
+      zone_redundant    = true
```
Comment (pasqualedevita) [inline infra/resources/_modules/cosmos/cosmos_account.tf:29]:
```suggestion
      zone_redundant    = false
```

Comment (pasqualedevita) [reply infra/resources/_modules/cosmos/cosmos_account.tf:29]: additional_geo_locations can be without zone redundancy

File: infra/resources/_modules/cosmos/cosmos_account.tf (hunk 1 above)
Comment (pasqualedevita) [reply infra/resources/_modules/cosmos/cosmos_account.tf:12]: main_geo_location_zone_redundant must be with zone redundancy

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  ]
+
+  backup_continuous_enabled = true
+
+  is_virtual_network_filter_enabled = true
```
Comment (pasqualedevita) [inline infra/resources/_modules/cosmos/cosmos_account.tf:35]: we can set to false

File: infra/resources/_modules/functions_apps/app_service_plan_wallet_common.tf [hunk 4]
```diff
@@ -0,0 +1,16 @@
+resource "azurerm_app_service_plan" "app_service_plan_wallet_common" {
```
Comment (pasqualedevita) [inline infra/resources/_modules/functions_apps/app_service_plan_wallet_common.tf:1]: resource deprecated, use https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/service_plan

Comment (pasqualedevita) [reply infra/resources/_modules/functions_apps/app_service_plan_wallet_common.tf:1]: add also zone_balancing_enabled = true and min instances to 3

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: infra/resources/prod/main.tf [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "<= 3.97.1"
+    }
+
+  }
+
+  backend "azurerm" {
+    resource_group_name  = "terraform-state-rg"
+    storage_account_name = "tfappprodio"
```
Comment (pasqualedevita) [inline infra/resources/prod/main.tf:13]:
```suggestion
    storage_account_name = "tfinfprodio"
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: infra/resources/_modules/functions_apps/function_app_wallet.tf [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    "private_dns_zone_queue_ids" = []
+    "private_dns_zone_table_ids" = []
+    "queues"                     = [],
+    "containers"                 = [],
+    "blobs_retention_days"       = 0,
+  }
+
+  subnet_id = var.subnet_id
+
+  allowed_subnets = [
```
Comment (pasqualedevita) [inline infra/resources/_modules/functions_apps/function_app_wallet.tf:35]: using fn apps and app service in italy north we need to use private endpoints

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: infra/resources/_modules/functions_apps/locals.tf [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      FETCH_KEEPALIVE_ENABLED             = "true"
+      FETCH_KEEPALIVE_SOCKET_ACTIVE_TTL   = "110000"
+      FETCH_KEEPALIVE_MAX_SOCKETS         = "40"
+      FETCH_KEEPALIVE_MAX_FREE_SOCKETS    = "10"
+      FETCH_KEEPALIVE_FREE_SOCKET_TIMEOUT = "30000"
+      FETCH_KEEPALIVE_TIMEOUT             = "60000"
+
+      // CosmosDB connection
+      COSMOSDB_URI           = var.cosmos_db.endpoint
+      COSMOSDB_KEY           = var.cosmos_db.primary_key
```
Comment (pasqualedevita) [inline infra/resources/_modules/functions_apps/locals.tf:29]: can we use managed identity to access database?

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: infra/resources/_modules/functions_apps/locals.tf (hunk 7 above)
Comment (gunzip) [reply infra/resources/_modules/functions_apps/locals.tf:29]: here's an example https://github.com/pagopa/io-fims/blob/main/infra/resources/prod/_modules/web_apps/app_service_openid_provider.tf#L36

Comment (pasqualedevita) [reply infra/resources/_modules/functions_apps/locals.tf:29]: correct we can define also custom roles for single container
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/cosmosdb_sql_role_assignment

Also you need to change application code to use managed identity authentication (cc @lucacavallaro)

**Review by pasqualedevita — CHANGES_REQUESTED** [review]
from policy iopitnwalletcdn01sa has public network access, we need to disable it

Screenshot 2024-05-14 at 15 41 36

**Comment by pasqualedevita** [comment]
Also is recommended to use a NAT Gateway for outgoing connections.
@Krusty93 we need to add a new NAT Gateway in Italy North https://github.com/pagopa/io-infra/tree/main/src/core/_modules/networking

File: infra/resources/_modules/functions_apps/locals.tf (hunk 7 above)
Comment (Krusty93) [reply infra/resources/_modules/functions_apps/locals.tf:29]: @christian-calabrese recently set up AI Search to access Cosmos DB. The same config could be used here as well

**Comment by Krusty93** [comment]
> Also is recommended to use a NAT Gateway for outgoing connections. @Krusty93 we need to add a new NAT Gateway in Italy North [pagopa/io-infra@`main`/src/core/_modules/networking](https://github.com/pagopa/io-infra/tree/main/src/core/_modules/networking?rgh-link-date=2024-05-14T13%3A46%3A57Z)

Is there any particular reason? Performance or security? Just asking, we can create it almost immediately

File: infra/resources/_modules/functions_apps/locals.tf (hunk 7 above)
Comment (christian-calabrese) [reply infra/resources/_modules/functions_apps/locals.tf:29]: If one of the [built in cosmosdb roles](https://learn.microsoft.com/en-us/azure/cosmos-db/how-to-setup-rbac#built-in-role-definitions) suits your case, you can also use one of them and reduce the scope to the database or collection level as done [here](https://github.com/pagopa/io-services-cms/blob/ebc186c93cf0966a2dc4f21ee138375dfb45f908/infra/_modules/ai_search/main.tf#L78C1-L84C2)

File: infra/resources/_modules/cdn/cdn.tf [hunk 8]
```diff
@@ -0,0 +1,29 @@
+module "wallet_cdn" {
+  source = "git::https://github.com/pagopa/terraform-azurerm-v3.git//cdn?ref=v8.12.2"
+
+  name                  = "wallet-cdn-01"
```
Comment (Krusty93) [inline infra/resources/_modules/cdn/cdn.tf:4]: Shouldn't be `io-p-wallet-cdn-01`?

File: infra/resources/_modules/cdn/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,7 @@
+output "storage_account" {
```
Comment (Krusty93) [inline infra/resources/_modules/cdn/outputs.tf:1]: I was thinking about an approach we took on other repos, and it was always group all NON sensitive properties under one value (only the necessary ones of course), keeping the sensitive ones in their own outputs. For example:

```hcl
output "storage_account" {
  value = {
    id = module.wallet_cdn.id
    name = module.wallet_cdn.name
  }
}

output "storage_account_primary_connection_string" {
  value = module.wallnet_cdn.primary_connection_string
  sensitive = true
}
```

What you've done is ok, I'm just saying

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 10]
```diff
@@ -0,0 +1,47 @@
+module "cosmos_account_wallet" {
+  source = "github.com/pagopa/terraform-azurerm-v3//cosmosdb_account?ref=v8.12.2"
+
+  name                = "${var.project}-cosmos-wallet"
```
Comment (Krusty93) [inline infra/resources/_modules/cosmos/cosmos_account.tf:4]:
```suggestion
  name                = "${var.project}-wallet-cosno-01"
```

[Source](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-abbreviations#databases)

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  offer_type = "Standard"
+  kind       = "GlobalDocumentDB"
+
+  main_geo_location_zone_redundant = false
+
+  enable_free_tier          = false
+  enable_automatic_failover = true
+
+  consistency_policy = {
+    consistency_level       = "Strong"
```
Comment (Krusty93) [inline infra/resources/_modules/cosmos/cosmos_account.tf:18]: Are we sure about this consistency level?

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  ]
+
+  backup_continuous_enabled = true
+
+  is_virtual_network_filter_enabled = true
+
+  ip_range = ""
+
+  # private endpoint
+  private_endpoint_sql_name           = "${var.project}-cosmos-wallet-sql-endpoint"
```
Comment (Krusty93) [inline infra/resources/_modules/cosmos/cosmos_account.tf:40]:
```suggestion
  private_endpoint_sql_name           = "${var.project}-wallet-sql-pep-01"
```

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  private_endpoint_enabled            = true
+  private_service_connection_sql_name = "${var.project}-cosmos-wallet-sql-endpoint"
```
Comment (Krusty93) [inline infra/resources/_modules/cosmos/cosmos_account.tf:42]:
```suggestion
  private_service_connection_sql_name = "${var.project}-wallet-sql-pep-01"
```

File: infra/resources/_modules/cdn/cdn.tf (hunk 8 above)
Comment (balanza) [reply infra/resources/_modules/cdn/cdn.tf:4]: the module has the `prefix` attribute that does the trick

File: infra/resources/_modules/cosmos/cosmos_account.tf [hunk 14]
```diff
@@ -0,0 +1,47 @@
+module "cosmos_account_wallet" {
```
Comment (Krusty93) [inline infra/resources/_modules/cosmos/cosmos_account.tf:1]: @pasqualedevita Could you please confirm we would like to use CMK?

@balanza In that case, we need to add a KeyVault reference using the `key_vault_key_id` property (already supported by this module) and reference the key. Please also note [these suggestions from HashiCopr](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/cosmosdb_account#key_vault_key_id)

File: infra/resources/_modules/networking/subnet_wallet.tf [hunk 15]
```diff
@@ -0,0 +1,24 @@
+module "subnet" {
+  source = "github.com/pagopa/terraform-azurerm-v3//subnet?ref=v8.12.2"
+
+  name                 = "${var.project}-wallet-snet-001"
```
Comment (Krusty93) [inline infra/resources/_modules/networking/subnet_wallet.tf:4]:
```suggestion
  name                 = "${var.project}-wallet-snet-01"
```

File: infra/resources/prod/locals.tf [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+  location           = "italynorth"
+  secondary_location = "germanywestcentral"
+
+  tags = {
+    CostCenter     = "TS310 - PAGAMENTI & SERVIZI"
+    CreatedBy      = "Terraform"
+    Environment    = "Prod"
+    Owner          = "IO"
+    ManagementTeam = "IO Wallet"
+    Source         = "https://github.com/pagopa/eudi-wallet-it-wallet-solution/blob/main/infra/resources/prod/italynorth"
```
Comment (Krusty93) [inline infra/resources/prod/locals.tf:23]:
```suggestion
    Source         = "https://github.com/pagopa/eudi-wallet-it-wallet-solution/blob/main/infra/resources/prod"
```

File: infra/resources/_modules/cosmos/cosmos_account.tf (hunk 10 above)
Comment (balanza) [reply infra/resources/_modules/cosmos/cosmos_account.tf]: done in https://github.com/pagopa/io-wallet/pull/56/commits/3f1ae7f83db0977b986e80e4ba6cbf4590b647b6

File: infra/resources/_modules/functions_apps/app_service_plan_wallet_common.tf (hunk 4 above)
Comment (balanza) [reply infra/resources/_modules/functions_apps/app_service_plan_wallet_common.tf]: Done in https://github.com/pagopa/io-wallet/pull/56/commits/9aca3808ab1fb50c4b5b792a2d6efa5b174ecd2b.

File: infra/resources/prod/main.tf (hunk 5 above)
Comment (balanza) [reply infra/resources/prod/main.tf:13]: Do we have a rationale? Afaik we are moving all the state files into `tfappprodio` cc @Krusty93

File: infra/resources/_modules/functions_apps/locals.tf (hunk 7 above)
Comment (balanza) [reply infra/resources/_modules/functions_apps/locals.tf]: Done in https://github.com/pagopa/io-wallet/pull/56/commits/a06686432f7f71d00738cc8e9e00e10a17b312d3

The `DocumentDB Account Contributor` should be right (see https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles/databases#documentdb-account-contributor).

**Comment by balanza** [comment]
> from policy iopitnwalletcdn01sa has public network access, we need to disable it

@pasqualedevita good point. Actually the storage is created by the CDN module which [hardcoded this option to true](https://github.com/pagopa/terraform-azurerm-v3/blob/c5e3aade01153d9dbd0bfb4f32bc6834fbded3c2/cdn/main.tf#L22-L23). 

We can update the module to accept an optional variable, of course. I wonde r, is it a problem for the other storages created by CDNs?

File: infra/resources/_modules/cosmos/alerts.tf [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    aggregation            = "Total"
+    operator               = "GreaterThan"
+    threshold              = 0
+    skip_metric_validation = false
+
+
+    dimension {
+      name     = "Region"
+      operator = "Include"
+      values   = ["North Italy"]
```
Comment (christian-calabrese) [inline infra/resources/_modules/cosmos/alerts.tf:26]: I think the region name is Italy North, if you'd agree I would change it to:
```suggestion
      values   = ["Italy North"]
```

**Comment by balanza** [comment]
@pasqualedevita @Krusty93 @christian-calabrese I propose to leave the following for a PR to be done next:
* NAT integration
* private endpoint for Azure function

Leaving them out, what's missing in this PR?

**Comment by pasqualedevita** [comment]
> > from policy iopitnwalletcdn01sa has public network access, we need to disable it
> 
> @pasqualedevita good point. Actually the storage is created by the CDN module which [hardcoded this option to true](https://github.com/pagopa/terraform-azurerm-v3/blob/c5e3aade01153d9dbd0bfb4f32bc6834fbded3c2/cdn/main.tf#L22-L23).
> 
> We can update the module to accept an optional variable, of course. I wonder, is it a problem for the other storage created by CDNs?

yes is a problem on all CDN, we can use an optional parameter with default value to private

**Comment by pasqualedevita** [comment]
> @pasqualedevita @Krusty93 @christian-calabrese I propose to leave the following for a PR to be done next:
> 
> * NAT integration
> * private endpoint for Azure function
> 
> Leaving them out, what's missing in this PR?

yes ok next PR

File: infra/resources/_modules/functions_apps/locals.tf (hunk 7 above)
Comment (pasqualedevita) [reply infra/resources/_modules/functions_apps/locals.tf]: yes `DocumentDB Account Contributor` is right.
You can use custom role to give more restrictive access (es. read-only a container)

File: infra/resources/prod/main.tf (hunk 5 above)
Comment (pasqualedevita) [reply infra/resources/prod/main.tf:13]: `inf` is for infrastructure where we can be a possible secrets in terraform state file

**Comment by pasqualedevita** [comment]
> > Also is recommended to use a NAT Gateway for outgoing connections. @Krusty93 we need to add a new NAT Gateway in Italy North [pagopa/io-infra@`main`/src/core/_modules/networking](https://github.com/pagopa/io-infra/tree/main/src/core/_modules/networking?rgh-link-date=2024-05-14T13%3A46%3A57Z)
> 
> Is there any particular reason? Performance or security? Just asking, we can create it almost immediately

Security: only if this function needs outgoing connections with static ip (es. a third party service require to whitelist our ips)
Performance: again if this function needs outgoing connections using NAT gateway should avoid SNAT exhaustion

File: infra/resources/_modules/cosmos/cosmos_account.tf (hunk 14 above)
Comment (pasqualedevita) [reply infra/resources/_modules/cosmos/cosmos_account.tf:1]: for me it's ok to use a key vault CMK, but be sure to set purge protection on keyvault at least 90 days

@pp-ps are u agree?

Comment (pp-ps) [reply infra/resources/_modules/cosmos/cosmos_account.tf:1]: Agreed, we will spread more information on CMKs in the following months.

**Comment by grausof** [comment]
Closed in favor of individual PRs on infra

**Pull request closed without merging**